src/winforms/toga_winforms/widgets/base.py

```diff
@@ -38,17 +39,26 @@ def set_enabled(self, value):
     ### APPLICATOR
 
     def set_bounds(self, x, y, width, height):
+        # If a widget is the top content of the window, it needs
+        # to be shifted down to accommodate the height of the
+        # toolbar
+        vertical_shift = 0
         if self.native:
+            try:
+                if isinstance(self.native.Parent, WinForms.Form):
```

Comment:
In general, I'd rather avoid `isinstance` checks - they mean you're checking for a specific class, rather than a useful property of a class. In this particular case, it means you're tying specific properties of the *window* to the implementation of the widget.

Would it be possible to establish if the widget is "top level" by checking `self.interface.parent is None`, and if so, asking the container for it's vertical shift? Most containers (split containers, scroll containers) will have a vertical shift of 0; windows can return a value based on whether they have a toolbar or not.

Reply:
If I check if `self.interface.parent is None`, several containers get to add vertical shift. In tutorial 2 they are table, scrollcontainer and splitcontainer, they don't seem to have a parent. Is it correct?

I added `if self.interface.window.content == self.interface:` instead of `isinstance` check, and it works :)

Reply:
Nice - `self.interface.window.content == self.interface` is a clever way to do that check :-)

I'm a little surprised at Table, but ScrollContainer and SplitContainer both make sense that they would have `parent is None` - because they're "top level containers" from the perspective of their underlying widget systems. 

What I was proposing was to say *all* top level containers have a vertical shift - but that in the case of ScrollContainer and SplitContainer (and probably OptionContainer too), that vertical shift was 0. And, rather than the base widget implementation knowing details about Window specifically, the base widget asks for the *container* to describe "how big is your vertical shift"? This can be a property with a value of 0 for SplitContainer; but Window would return a value matching the size of it's toolbar. 

Checking for Window specifically means we've carved out a special case for the Window container. At the moment, that's probably all we need. However, if we can abstract the idea - some container widgets need a vertical offset - then we don't need to spell out every possible special case in the future.

For example, it wouldn't surprise me if OptionContainer will need a vertical offset as well. It's a container whose content may need to be offset a little bit from the top of the container itself (to allow for the tabs/buttons to switch between options). Rather than carve out a second "special case" for OptionContainer, we can rely on the containers themselves to specify the offset. That offset is 0 in the case of most containers; but Window needs to account for the toolbar (which might not even be there), and OptionContainer needs to know the size of the tab area.

Does that make sense?